rpc/client.go

```diff
@@ -2,21 +2,62 @@ package rpc
 
 import (
 	"context"
-	"path/filepath"
+	"runtime"
 
 	"github.com/evergreen-ci/poplar"
 	"github.com/evergreen-ci/poplar/rpc/internal"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/queue"
 	"github.com/mongodb/grip"
 	"github.com/mongodb/grip/message"
 	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 )
 
 func UploadReport(ctx context.Context, report *poplar.Report, cc *grpc.ClientConn, dryRun bool) error {
+	if err := convertAndUploadArtifacts(ctx, report, dryRun); err != nil {
+		return errors.Wrap(err, "problem uploading tests for report")
+	}
 	return errors.Wrap(uploadTests(ctx, internal.NewCedarPerformanceMetricsClient(cc), report, report.Tests, dryRun),
 		"problem uploading tests for report")
 }
 
+func convertAndUploadArtifacts(ctx context.Context, report *poplar.Report, dryRun bool) error {
+	jobQueue := queue.NewLocalUnordered(runtime.NumCPU())
+	if err := jobQueue.Start(ctx); err != nil {
+		return errors.Wrap(err, "problem starting artifact upload queue")
+	}
+
+	queue := make([]poplar.Test, 0)
+	for _, test := range report.Tests {
+		queue = append(queue, test)
+	}
+
+	for len(queue) != 0 {
```

Comment:
I'd use range but only the first element.

Reply:
breadth-first-search :)